### ISSUE TITLE
Bitmex: TBP Logic 

### ISSUE THREAD

**Issue by asfbhero** [ctubio/Krypto-trading-bot#986]
I can’t get tbp to work as it should. Now that I’ve got things working in terms of trading it’s now time to work on risk management. 

As of now it looks like the bot doesn’t understand if tbp held should be applied to the bid side or the ask side. I imagine this could be fixed by having the bot query the bitmex api and see if the net position is long or short, then hold that side.

Coding this seems a bit over my head but I’ll give it a shot, @ctubio can you point me in the right direction and I’ll poke at it after work tonight.

**Comment by Camille92** [comment]
Are you referring to https://github.com/ctubio/Krypto-trading-bot/issues/976 ?

If not TBP is calculated with the ewma !
Anyways I don't think net position is a good help here.

**Comment by asfbhero** [comment]
I don’t think it’s what was discussed there. 

Say I’m using manual mode with 1 btc in margin. I want the bot to stop buying when I’ve spent 50% of my margin on buys and stop selling when I’ve used 50% of my margin on sells.


The problem is that since the margin is in bitcoin, the bot only sees that Bitcoin held has gone down, regardless if I’ve used that bitcoin margin to accumulate long or short positions in bitcoin.

If I’m long eos/xbt it’s fine, because I’m net short bitcoin, having used 50% of my margin buying eos and selling xbt, so it will set the bid to tbp held.

But if I’m net short eos/btc, having used 50% of my margin buying btc and selling eos, it will still see that I have 50% of the bitcoin in my margin account, and it will set the bid to tbp held.

So no matter if you have too many Buys or too many Sells, the bot will only stop buying, and never stop selling.

This is different than the never stopping buying or selling caused by not knowing what’s a ping and what’s a pong which I resolved by setting the bot to respect trades/s no matter if the trade is a ping or a pong.

**Comment by MiningInMySleep** [comment]
Running into a possibly related issue where the bot, upon running out of margin it won't recognize that and proceeds to spam orders (which get "insufficient margin" error) until IP banned.

**Comment by asfbhero** [comment]
Yeah, it’s related. I haven’t really had time to play with it since I’m in the middle of a big project at work.

My only solution has been to come up with a new bitmex specific strategy that trades less and with bigger sizing. Though this obviously isn’t and ideal strategy with such high rebates available.

**Comment by ctubio** [comment]
lemme reread all comments slowly again to see if i can understand it

**Comment by jakeadelman** [comment]
I have this issue as well. Seems to me that it is because the TBHHeld logic is hardcoded to stop buying when tbpheld, instead of 1. fetching open positions 2. stop sending orders to the side that will reduce exposure

```
void applyTotalBasePosition() {
        if (wallet.base.total < wallet.target.targetBasePosition - wallet.target.positionDivergence) {
          quotes.ask.clear(QuoteState::TBPHeld);
          if (!quotes.bid.empty() and qp.aggressivePositionRebalancing != APR::Off) {
            sideAPR = SideAPR::Buy;
            if (!qp.buySizeMax)
              quotes.bid.size = fmin(
                qp.aprMultiplier * quotes.bid.size,
                K.gateway->margin == Future::Spot
                  ? wallet.target.targetBasePosition - wallet.base.total
                  : (K.gateway->margin == Future::Inverse
                      ? (wallet.target.targetBasePosition - wallet.base.total) * quotes.bid.price
                      : (wallet.target.targetBasePosition - wallet.base.total) / quotes.bid.price
                  )
              );
          }
        }
        else if (wallet.base.total >= wallet.target.targetBasePosition + wallet.target.positionDivergence) {
          quotes.bid.clear(QuoteState::TBPHeld);
          if (!quotes.ask.empty() and qp.aggressivePositionRebalancing != APR::Off) {
            sideAPR = SideAPR::Sell;
            if (!qp.sellSizeMax)
              quotes.ask.size = fmin(
                qp.aprMultiplier * quotes.ask.size,
                K.gateway->margin == Future::Spot
                  ? wallet.base.total - wallet.target.targetBasePosition
                  : (K.gateway->margin == Future::Inverse
                      ? (wallet.base.total - wallet.target.targetBasePosition) * quotes.ask.price
                      : (wallet.base.total - wallet.target.targetBasePosition) / quotes.ask.price
                  )
              );
          }
        }
        else sideAPR = SideAPR::Off;
      };
```

This looks like the function to me. In order to get the required info to the function we need to edit the bitmex api gateway I think. We could add in the url https://www.bitmex.com/api/v1/positions/ and query the open positions on the call of applyTotalBasePosition() function.
